### PR TITLE
Implement the batch user creation service

### LINKED ISSUE #456: Batch user creation service
## Overview 
Group managers may be allowed to create a set of users in batch (for a contest, ...). See [designs](https://france-ioi.github.io/algorea-designs/01.Groups_02.Composition_01.Dialogs_02.Create-batch-accounts) for the typical UI.

On `POST /user_batches/`

## Input
* group_prefix: required
* custom_prefix : required string matching `/^[a-z0-9-]{2,14}$/`
* subgroups: non-empty list of objects having:
   * group_id: required
   * count: required >0
* postfix_length: non-optional, min:3, max:29, 
* password_length: non-optional, min:6, max:50

## Actions
* create a user batch entry with the authenticated user as creator (should not fail if it is not a duplicate and FK exists)
* call the login module to create sum of subgroups.count accounts (see [1](https://github.com/France-ioi/login-module-client/blob/master/src/PlatformApi/AccountsManager.php#L19) with as param:
  - prefix: "{group_prefix}\_{custom_prefix}\_"
  - amount: the sum of subgroups.count
  - postfix_length:
  - password_length
  - login_fixed: true
  - language": authenticated user's `default_language`
* create the corresponding users in our db
* put the subgroup[0].count first users in subgroup[0].group_id, and so on ..., "auto-approve" all approvals required by each group

## Output

on success: 
```json
data: [
    { 
      "group_id": 567,
      "users": [
         {"user_id":123, "login":"...", "password":"..."},
         ...
      ]
    }, 
    ...
  ]
```

## Validations
* The authenticated user is manager of the group (or its ancestors) of the group_prefix with at least can_manage:memberships.
* The subgroup.group_id are descendant of the input.group_id or is the group_id itself
* the prefix is `allowed`
*  32^postfix_length > 2* sum of subgroups.count  (to prevent being unable to generate new login)
* sum of subgroups.count + sum of existing batches under the same group_prefix is not > `max_users` of the prefix.

### PR DESCRIPTION
**Pull request by zenovich**
Fixes #456

### CODE REVIEW

File: app/api/groups/create_user_batch.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+					"watch_approved_at":              watchApprovedAt},
+			)
+
+			usersToCreate = append(usersToCreate, map[string]interface{}{
+				"temp_user":     0,
+				"registered_at": database.Now(),
+				"group_id":      userGroupID,
+				"login_id":      createdUser.ID,
+				"login":         createdUser.Login,
+				"creator_id":    user.GroupID,
```
Comment (smadbe) [inline app/api/groups/create_user_batch.go]: default_language should probably be set as well

File: app/api/groups/create_user_batch.robustness.feature [hunk 2]
```diff
@@ -0,0 +1,304 @@
+Feature: Create a user batch - robustness
+  Background:
+  Background:
```
Comment (smadbe) [inline app/api/groups/create_user_batch.robustness.feature]: x2?

File: app/api/groups/create_user_batch.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+//   * creates new users in the login module,
+//
+//   * inserts the created users into the `users` table,
+//
+//   * adds the created users into groups specified as `subgroups[...].group_id` giving all the required approvals.
+//
+//
+//   Restrictions:
+//
+//   * The authenticated user (or one of his group ancestors) should be a manager of the group linked to the `group_prefix`
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/groups/create_user_batch.go]: fixed

File: app/api/groups/create_user_batch.go (hunk 1 above)
Comment (zenovich) [reply app/api/groups/create_user_batch.go]: fixed

File: app/api/groups/create_user_batch.robustness.feature (hunk 2 above)
Comment (zenovich) [reply app/api/groups/create_user_batch.robustness.feature]: fixed

File: app/api/groups/create_user_batch.feature [hunk 4]
```diff
@@ -98,10 +98,10 @@ Feature: Create a user batch
...
       | test         | custom        | 2    | 21         | 2019-07-16 22:02:28 |
     And the table "users" should be:
-      | group_id            | latest_login_at | latest_activity_at | temp_user | registered_at       | login_id  | login           | email | first_name  | last_name |
-      | 21                  | null            | null               | 0         | null                | null      | owner           | null  | Jean-Michel | Blanquer  |
-      | 5577006791947779410 | null            | null               | 0         | 2019-07-16 22:02:28 | 100000029 | test_custom_jzk | null  | null        | null      |
-      | 8674665223082153551 | null            | null               | 0         | 2019-07-16 22:02:28 | 100000030 | test_custom_ctc | null  | null        | null      |
+      | group_id            | latest_login_at | latest_activity_at | temp_user | registered_at       | login_id  | login           | default_language | email | first_name  | last_name |
+      | 21                  | null            | null               | 0         | null                | null      | owner           | fr               | null  | Jean-Michel | Blanquer  |
```
Comment (smadbe) [inline app/api/groups/create_user_batch.feature]: as "fr" is the default anyway, probably that's not a good choice to demonstrate that it is copied from the creator

Comment (zenovich) [reply app/api/groups/create_user_batch.feature]: fixed

**Pull request merged**